node/lib/util/submodule_config_util.js

```diff
@@ -70,6 +78,106 @@ const CONFIG_FILE_NAME = "config";
  */
 exports.modulesFileName = ".gitmodules";
 
+/**
+ * Remove the first found configuration entry for the submodule having the
+ * specified `submoduleName` in the config file in the specified `repoPath`; do
+ * nothing if no entry is found.
+ *
+ * @param {String} repoPath
+ * @param {String} submoduleName
+ */
+exports.clearSubmoduleConfigEntry =
+                                  co.wrap(function *(repoPath, submoduleName) {
+    assert.isString(repoPath);
+    assert.isString(submoduleName);
+
+    // Using a very stupid algorithm here to find and remove the submodule
+    // entry.  This logic could be smarter (maybe use regexes) and more
+    // efficition (stream in and out).
+
+    const configPath = path.join(repoPath, "config");
+    const configText = fs.readFileSync(configPath, "utf8");
+    const configLines = configText.split("\n");
+    const newConfigLines = [];
+    const searchString = "[submodule \"" + submoduleName + "\"]";
+
+    let found = false;
+    let inSubmoduleConfig = false;
+
+    // Loop through the file and push, onto 'newConfigLines' any lines that
+    // aren't part of the bad submodule section.
+
+    for (let i = 0; i < configLines.length; ++i) {
+        let line = configLines[i];
+        if (!found && !inSubmoduleConfig && line === searchString) {
+            inSubmoduleConfig = true;
+            found = true;
+        }
+        else if (inSubmoduleConfig) {
+            // If we see a line starting with "[" while we're in the submodule
+            // section, we can get out of it.
+
+            if (0 !== line.length && line[0] === "[") {
+                inSubmoduleConfig = false;
+            }
+        }
+        if (!inSubmoduleConfig) {
+            newConfigLines.push(line);
+        }
+    }
+
+    // If we didn't find the submodule, don't write the data back out.
+
+    if (found) {
+        newConfigLines.push("");  // one more new line
+        fs.writeFileSync(configPath, newConfigLines.join("\n"));
+    }
+
+    // Silence warning about no yield statement.
+    yield (Promise.resolve());
```

Comment:
why not just use the async fs-promise funcs to read the file?

Reply:
I use the synchronous methods when mutating the config file to avoid race conditions.

Reply:
(will add a note explaining this)